src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs

```diff
@@ -274,6 +258,8 @@ private Exception DownloadResource(IHostEnvironment env, IChannel ch, WebClient
                 using (var ws = fh.CreateWriteStream())
                 {
                     var headers = webClient.ResponseHeaders.GetValues("Content-Length");
+                    if (IsRedirectToDefaultPage(uri))
```

Comment:
This still does not answer the question as to why we have invalid aka.ms urls. Which urls are those?

Reply:
These URLs are those that are outputted as absoluteUrl here:

Reply:
That means they are not aka.ms urls. Is that correct? If so, why would they be subject to redirection?

---
In reply to: [394132286](https://github.com/dotnet/machinelearning/pull/4943#discussion_r394132286) [](ancestors = 394132286)

Reply:
Not always. In PR #4838, I am trying to use `EnsureResourceAsync` instead of `DownloadIfNotExists` here, to use the async nature of `EnsureResourceAsync` and fix hanging issues with some of our hanging tests :

https://github.com/dotnet/machinelearning/blob/d849ba4c7831586018d5b2cca6bdd2b1fa228ddf/test/Microsoft.ML.AutoML.Tests/DatasetUtil.cs#L40-L48

I found that when I use `EnsureResourceAsync` here, and ran the test `AutoFitRegressionTest`, it attempted to download from the url `https://aka.ms/mlnet-resources/https://raw.githubusercontent.com/dotnet/machinelearning/e78971ea6fd736038b4c355b840e5cbabae8cb55/test/data/generated_regression_dataset.csv%22`, which is obviously false. This link was subject to redirection.

Reply:
**Why have an aka.ms test:**
When creating a new component that uses ResourceManagerUtils' `EnsureResourceAsync()`, or when modifying a component's download path, it's easy to get the path wrong. Detection difficulty is exacerbated as incorrect aka.ms URLs silently return a webpage instead of a 404.

The previous check for this was broken. The method I recommended was checking the final URI (for uri.host=="www.microsoft.com" && uri.ToString().Length < somelimit). 

Having a 301/302 is an ok response for URIs. So may not be strict enough.

Perhaps check for: return code, & final host. Perhaps restricting the check to only aka.ms links (in case someone want to use `ResourceManagerUtils` for non-aka.ms links).

If we want more strict checking we could verify the filesize and/or file hash. We would also need versioned/immutable resource files in our stores (or old ML․NET versions will fail validation w/ if a file is updated).

**Where to store resource files:**
For large files, all of our download URLs should be aka.ms links pointing to a CDN (and CDN retrieves from the actual location). The aka.ms link give us forward-compatibility as we can point these to new locations should we modify our file hosting locations. The CDN gives us good worldwide download speeds.

Storage locations:
* Tiny unit test sized files -- in repo, use as a local file when possible
* Medium -- alternative repo w/ pinned commit (not master), downloaded by `ResourceManagerUtils`
* Large and above -- aka.ms link to CDN, downloaded by `ResourceManagerUtils`

The URLs [[1](https://raw.githubusercontent.com/dotnet/machinelearning/e78971ea6fd736038b4c355b840e5cbabae8cb55/test/data/generated_regression_dataset.csv), [2](https://raw.githubusercontent.com/dotnet/machinelearning/e78971ea6fd736038b4c355b840e5cbabae8cb55/test/data/generated_regression_dataset.csv")] are linking into this repo. Generally I don't think there's much reason for grabbing from the same repo using a web call; this was likely created when the AutoML code was in a different repo.

Reply:
I think you mean PR #4938. 

So, the first decision to make is whether EnsureResourceAsync is exclusively for downloading from aka.ms urls or not. 

If it is exclusively for aka.ms urls, then it will always prepend aka.ms, and your usage of this function in PR #4938 is wrong.

If however we wish to treat this as a generic download function, then you cannot assume that it will always redirect back to microsoft.com. In this case checking for http codes is the only reliable thing to do. 

Justin is right that you shouldn't be using web requests to download files that are already in the repo. If we also adopt his suggestion that all resources ML.NET uses should be abstracted behind aka.ms, then it is okay to check for final redirection going back to microsoft. And if we do detect such a condition that means our usage is wrong or the aka.ms entry needs to be updated and in this case we should be throwing an exception. 

So please figure out what else this function is needed for and whether we are willing to adopt the policy that we are going to use aka.ms urls for everything. 

---
In reply to: [394139313](https://github.com/dotnet/machinelearning/pull/4943#discussion_r394139313) [](ancestors = 394139313)

Reply:
We could allow the calling function to verify the file, and we would need to return the needed info to the calling function to let it make a decision. For instance, the final URI.

Though it's a bit more complicated as we'd then be outside of the mutex. We'll be causing a race condition if we allow the calling function to then delete the file.
  
thread 1: calls `EnsureResourceAsync()`, gets mutex
thread 2: calls `EnsureResourceAsync()`, waiting on mutex

thread 1: downloads file, unlocks mutex, returns to calling function
thread 2: grabs mutex, sees that file exists, unlocks mutex, returns to its calling function

thread 1: checks and deletes file
thread 2: opens file, ...{{dies}}...

Perhaps we could do dependency injection style of sending an optional lambda function to `EnsureResourceAsync()` which verifies the file while within the mutex?

Reply:
Yeah, I think this would be a bit of overkill. I would rather let the calling function do the higher level verification. 

---
In reply to: [394608927](https://github.com/dotnet/machinelearning/pull/4943#discussion_r394608927) [](ancestors = 394608927)

Reply:
Verification done properly (no race conditions) from the calling function would involve another named mutex, and much of the steps provided by `EnsureResourceAsync()`.

I do agree this is bit of overkill. I'd recommend the simple route of having `EnsureResourceAsync()` check for bad aka.ms links and leave the verification at that.

Reply:
@mstfbl Can you please verify if this code path always will be used only for aka.ms links? Only in that case I think we can explicitly check for microsoft.com as the redirected uri for failure. 

---
In reply to: [394646902](https://github.com/dotnet/machinelearning/pull/4943#discussion_r394646902) [](ancestors = 394646902)

Reply:
@harishsk `EnsureResourceAsync()` will be used only for aka.ms links. The AutoML test datasets that are currently downloaded with `DownloadIfNotExists` from earlier commits exist in the ML.NET repo, so I see now there is no need to use `EnsureResourceAsync()` for them.
In this case, I'll add back the explicit check for https://microsoft.com/ as the redirected url for failure.

Reply:
Are you sure? Can you please take a look at line 39 in ResourcemanagerUtils.cs?
It appears that MlNetResourcesUrl can be overridden by an environment variable. How would it work if somebody ran this code with a different url set in the environment?
What is the use case for that?

Reply:
Noted. The purpose of IsRedirectToDefaultPage(string url) is to distinguish the redirects that can happen from faulty aka.ms URLs. I added a check where this redirect check would only apply to aka.ms URLs.